File: fe/src/main/java/org/apache/doris/catalog/Column.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.catalog;

import org.apache.doris.alter.SchemaChangeHandler;
import org.apache.doris.analysis.Expr;
import org.apache.doris.common.CaseSensibility;
import org.apache.doris.common.DdlException;
import org.apache.doris.common.FeMetaVersion;
import org.apache.doris.common.io.Text;
import org.apache.doris.common.io.Writable;
import org.apache.doris.persist.gson.GsonUtils;
import org.apache.doris.thrift.TColumn;
import org.apache.doris.thrift.TColumnType;

import com.google.common.base.Strings;
import com.google.gson.annotations.SerializedName;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;

/**
 * This class represents the column-related metadata.
 */
public class Column implements Writable {
    private static final Logger LOG = LogManager.getLogger(Column.class);
    @SerializedName(value = "name")
    private String name;
    @SerializedName(value = "type")
    private Type type;
    // column is key: aggregate type is null
    // column is not key and has no aggregate type: aggregate type is none
    // column is not key and has aggregate type: aggregate type is name of aggregate function.
    @SerializedName(value = "aggregationType")
    private AggregateType aggregationType;

    // if isAggregationTypeImplicit is true, the actual aggregation type will not be shown in show create table
    // the key type of table is duplicate or unique: the isAggregationTypeImplicit of value columns are true
    // other cases: the isAggregationTypeImplicit is false
    @SerializedName(value = "isAggregationTypeImplicit")
    private boolean isAggregationTypeImplicit;
    @SerializedName(value = "isKey")
    private boolean isKey;
    @SerializedName(value = "isAllowNull")
    private boolean isAllowNull;
    @SerializedName(value = "defaultValue")
    private String defaultValue;
    @SerializedName(value = "comment")
    private String comment;
    @SerializedName(value = "stats")
    private ColumnStats stats;     // cardinality and selectivity etc.
    private Expr defineExpr; // use to define column in materialize view

    public Column() {
        this.name = "";
        this.type = Type.NULL;
        this.isAggregationTypeImplicit = false;
        this.isKey = false;
        this.stats = new ColumnStats();
    }

    public Column(String name, PrimitiveType dataType) {
        this(name, ScalarType.createType(dataType), false, null, false, null, "");
    }

    public Column(String name, PrimitiveType dataType, boolean isAllowNull) {
        this(name, ScalarType.createType(dataType), false, null, isAllowNull, null, "");
    }

    public Column(String name, Type type) {
        this(name, type, false, null, false, null, "");
    }

    public Column(String name, Type type, boolean isKey, AggregateType aggregateType, String defaultValue,
                  String comment) {
        this(name, type, isKey, aggregateType, false, defaultValue, comment);
    }

    public Column(String name, Type type, boolean isKey, AggregateType aggregateType, boolean isAllowNull,
                  String defaultValue, String comment) {
        this.name = name;
        if (this.name == null) {
            this.name = "";
        }

        this.type = type;
        if (this.type == null) {
            this.type = Type.NULL;
        }

        this.aggregationType = aggregateType;
        this.isAggregationTypeImplicit = false;
        this.isKey = isKey;
        this.isAllowNull = isAllowNull;
        this.defaultValue = defaultValue;
        this.comment = comment;

        this.stats = new ColumnStats();
    }

    public Column(Column column) {
        this.name = column.getName();
        this.type = column.type;
        this.aggregationType = column.getAggregationType();
        this.isAggregationTypeImplicit = column.isAggregationTypeImplicit();
        this.isKey = column.isKey();
        this.isAllowNull = column.isAllowNull();
        this.defaultValue = column.getDefaultValue();
        this.comment = column.getComment();
        this.stats = column.getStats();
    }

    public void setName(String newName) {
        this.name = newName;
    }

    public String getName() {
        return this.name;
    }

    public String getNameWithoutPrefix(String prefix) {
        if (isNameWithPrefix(prefix)) {
            return name.substring(prefix.length());
        }
        return name;
    }

    public boolean isNameWithPrefix(String prefix) {
        return this.name.startsWith(prefix);
    }

    public void setIsKey(boolean isKey) {
        this.isKey = isKey;
    }

    public boolean isKey() {
        return this.isKey;
    }

    public PrimitiveType getDataType() { return type.getPrimitiveType(); }

    public Type getType() { return ScalarType.createType(type.getPrimitiveType()); }

    public void setType(Type type) {
        this.type = type;
    }

    public Type getOriginType() { return type; }

    public int getStrLen() { return ((ScalarType) type).getLength(); }
    public int getPrecision() { return ((ScalarType) type).getScalarPrecision(); }
    public int getScale() { return ((ScalarType) type).getScalarScale(); }

    public AggregateType getAggregationType() {
        return this.aggregationType;
    }

    public boolean isAggregated() {
        return aggregationType != null && aggregationType != AggregateType.NONE;
    }

    public boolean isAggregationTypeImplicit() {
        return this.isAggregationTypeImplicit;
    }

    public void setAggregationType(AggregateType aggregationType, boolean isAggregationTypeImplicit) {
        this.aggregationType = aggregationType;
        this.isAggregationTypeImplicit = isAggregationTypeImplicit;
    }

    public void setAggregationTypeImplicit(boolean isAggregationTypeImplicit) {
        this.isAggregationTypeImplicit = isAggregationTypeImplicit;
    }

    public boolean isAllowNull() {
        return isAllowNull;
    }

    public void setIsAllowNull(boolean isAllowNull) {
        this.isAllowNull = isAllowNull;
    }

    public String getDefaultValue() {
        return this.defaultValue;
    }

    public void setStats(ColumnStats stats) {
        this.stats = stats;
    }

    public ColumnStats getStats() {
        return this.stats;
    }

    public void setComment(String comment) {
        this.comment = comment;
    }

    public String getComment() {
        return comment;
    }

    public int getOlapColumnIndexSize() {
        PrimitiveType type = this.getDataType();
        if (type == PrimitiveType.CHAR) {
            return getStrLen();
        } else {
            return type.getOlapColumnIndexSize();
        }
    }

    public TColumn toThrift() {
        TColumn tColumn = new TColumn();
        tColumn.setColumn_name(this.name);

        TColumnType tColumnType = new TColumnType();
        tColumnType.setType(this.getDataType().toThrift());
        tColumnType.setLen(this.getStrLen());
        tColumnType.setPrecision(this.getPrecision());
        tColumnType.setScale(this.getScale());

        tColumnType.setIndex_len(this.getOlapColumnIndexSize());

        tColumn.setColumn_type(tColumnType);
        if (null != this.aggregationType) {
            tColumn.setAggregation_type(this.aggregationType.toThrift());
        }
        tColumn.setIs_key(this.isKey);
        tColumn.setIs_allow_null(this.isAllowNull);
        tColumn.setDefault_value(this.defaultValue);
        if (this.defineExpr != null) {
            tColumn.setDefine_expr(this.defineExpr.treeToThrift());
        }
        return tColumn;
    }

    public void checkSchemaChangeAllowed(Column other) throws DdlException {
        if (Strings.isNullOrEmpty(other.name)) {
            throw new DdlException("Dest column name is empty");
        }

        if (!ColumnType.isSchemaChangeAllowed(type, other.type)) {
            throw new DdlException("Can not change " + getDataType() + " to " + other.getDataType());
        }

        if (this.aggregationType != other.aggregationType) {
            throw new DdlException("Can not change aggregation type");
        }

        if (this.isAllowNull && !other.isAllowNull) {
            throw new DdlException("Can not change from nullable to non-nullable");
        }

        if (this.getDefaultValue() == null) {
            if (other.getDefaultValue() != null) {
                throw new DdlException("Can not change default value");
            }
        } else {
            if (!this.getDefaultValue().equals(other.getDefaultValue())) {
                throw new DdlException("Can not change default value");
            }
        }

        if ((getDataType() == PrimitiveType.VARCHAR && other.getDataType() == PrimitiveType.VARCHAR)
                || (getDataType() == PrimitiveType.CHAR && other.getDataType() == PrimitiveType.VARCHAR)
                || (getDataType() == PrimitiveType.CHAR && other.getDataType() == PrimitiveType.CHAR)) {
            if (getStrLen() > other.getStrLen()) {
                throw new DdlException("Cannot shorten string length");
            }
        }

        // now we support convert decimal to varchar type
        if ((getDataType() == PrimitiveType.DECIMAL && other.getDataType() == PrimitiveType.VARCHAR)
                || (getDataType() == PrimitiveType.DECIMALV2 && other.getDataType() == PrimitiveType.VARCHAR)) {
            return;
        }

        if (this.getPrecision() != other.getPrecision()) {
            throw new DdlException("Cannot change precision");
        }

        if (this.getScale() != other.getScale()) {
            throw new DdlException("Cannot change scale");
        }
    }

    public boolean nameEquals(String otherColName, boolean ignorePrefix) {
        if (CaseSensibility.COLUMN.getCaseSensibility()) {
            if (!ignorePrefix) {
                return name.equals(otherColName);
            } else {
                return removeNamePrefix(name).equals(removeNamePrefix(otherColName));
            }
        } else {
            if (!ignorePrefix) {
                return name.equalsIgnoreCase(otherColName);
            } else {
                return removeNamePrefix(name).equalsIgnoreCase(removeNamePrefix(otherColName));
            }
        }
    }

    public static String removeNamePrefix(String colName) {
        if (colName.startsWith(SchemaChangeHandler.SHADOW_NAME_PRFIX)) {
            return colName.substring(SchemaChangeHandler.SHADOW_NAME_PRFIX.length());
        }
        return colName;
    }

    public Expr getDefineExpr() {
        return defineExpr;
    }

    public void setDefineExpr(Expr expr) {
        defineExpr = expr;
    }

    public String toSql() {
        StringBuilder sb = new StringBuilder();
        sb.append("`").append(name).append("` ");
        String typeStr = type.toSql();
        sb.append(typeStr).append(" ");
        if (aggregationType != null && aggregationType != AggregateType.NONE && !isAggregationTypeImplicit) {
            sb.append(aggregationType.name()).append(" ");
        }
        if (isAllowNull) {
            sb.append("NULL ");
        } else {
            sb.append("NOT NULL ");
        }
        if (defaultValue != null && getDataType() != PrimitiveType.HLL && getDataType() != PrimitiveType.BITMAP) {
            sb.append("DEFAULT \"").append(defaultValue).append("\" ");
        }
        sb.append("COMMENT \"").append(comment).append("\"");

        return sb.toString();
    }

    @Override
    public String toString() {
        return toSql();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        }
        if (!(obj instanceof Column)) {
            return false;
        }

        Column other = (Column) obj;

        if (!this.name.equalsIgnoreCase(other.getName())) {
            return false;
        }
        if (this.getDataType() != other.getDataType()) {
            return false;
        }
        if (this.aggregationType != other.getAggregationType()) {
            return false;
        }
        if (this.isAggregationTypeImplicit != other.isAggregationTypeImplicit()) {
            return false;
        }
        if (this.isKey != other.isKey()) {
            return false;
        }
        if (this.isAllowNull != other.isAllowNull) {
            return false;
        }
        if (this.getDefaultValue() == null) {
            if (other.getDefaultValue() != null) {
                return false;
            }
        } else {
            if (!this.getDefaultValue().equals(other.getDefaultValue())) {
                return false;
            }
        }

        if (this.getStrLen() != other.getStrLen()) {
            return false;
        }
        if (this.getPrecision() != other.getPrecision()) {
            return false;
        }
        if (this.getScale() != other.getScale()) {
            return false;
        }

        if (!comment.equals(other.getComment())) {
            return false;
        }

        return true;
    }

    @Override
    public void write(DataOutput out) throws IOException {
        String json = GsonUtils.GSON.toJson(this);
        Text.writeString(out, json);
    }

    private void readFields(DataInput in) throws IOException {
        name = Text.readString(in);
        type = ColumnType.read(in);
        boolean notNull = in.readBoolean();
        if (notNull) {
            aggregationType = AggregateType.valueOf(Text.readString(in));

            if (Catalog.getCurrentCatalogJournalVersion() >= FeMetaVersion.VERSION_30) {
                isAggregationTypeImplicit = in.readBoolean();
            } else {
                isAggregationTypeImplicit = false;
            }
        }

        if (Catalog.getCurrentCatalogJournalVersion() >= FeMetaVersion.VERSION_30) {
            isKey = in.readBoolean();
        } else {
            isKey = (aggregationType == null);
        }

        if (Catalog.getCurrentCatalogJournalVersion() >= FeMetaVersion.VERSION_22) {
            isAllowNull = in.readBoolean();
        } else {
            isAllowNull = false;
        }

        notNull = in.readBoolean();
        if (notNull) {
            defaultValue = Text.readString(in);
        }
        stats = ColumnStats.read(in);

        if (Catalog.getCurrentCatalogJournalVersion() >= FeMetaVersion.VERSION_10) {
            comment = Text.readString(in);
        } else {
            comment = "";
        }
    }

    public static Column read(DataInput in) throws IOException {
        if (Catalog.getCurrentCatalogJournalVersion() < FeMetaVersion.VERSION_86) {
            Column column = new Column();
            column.readFields(in);
            return column;
        } else {
            String json = Text.readString(in);
            return GsonUtils.GSON.fromJson(json, Column.class);
        }
    }
}

File: be/src/olap/memory/column.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "olap/memory/column.h"

#include "olap/memory/typed_column_reader.h"
#include "olap/memory/typed_column_writer.h"

namespace doris {
namespace memory {

Column::Column(const ColumnSchema& cs, ColumnType storage_type, uint64_t version)
        : _cs(cs), _storage_type(storage_type), _base_idx(0) {
    _base.reserve(BASE_CAPACITY_MIN_STEP_SIZE);
    _versions.reserve(VERSION_CAPACITY_STEP_SIZE);
    _versions.emplace_back(version);
}

Column::Column(const Column& rhs, size_t new_base_capacity, size_t new_version_capacity)
        : _cs(rhs._cs), _storage_type(rhs._storage_type), _base_idx(rhs._base_idx) {
    _base.reserve(std::max(new_base_capacity, rhs._base.capacity()));
    _base.resize(rhs._base.size());
    for (size_t i = 0; i < _base.size(); i++) {
        _base[i] = rhs._base[i];
    }
    _versions.reserve(std::max(new_version_capacity, rhs._versions.capacity()));
    _versions.resize(rhs._versions.size());
    for (size_t i = 0; i < _versions.size(); i++) {
        _versions[i] = rhs._versions[i];
    }
}

size_t Column::memory() const {
    size_t bs = _base.size();
    size_t ds = _versions.size();
    size_t base_memory = 0;
    for (size_t i = 0; i < bs; i++) {
        base_memory += _base[i]->memory();
    }
    size_t delta_memory = 0;
    for (size_t i = 0; i < ds; i++) {
        if (_versions[i].delta) {
            delta_memory += _versions[i].delta->memory();
        }
    }
    return base_memory + delta_memory;
}

std::string Column::debug_string() const {
    return StringPrintf("Column(%s version=%zu)", _cs.debug_string().c_str(),
                        _versions.back().version);
}

Status Column::capture_version(uint64_t version, std::vector<ColumnDelta*>* deltas,
                               uint64_t* real_version) const {
    uint64_t base_version = _versions[_base_idx].version;
    *real_version = base_version;
    if (version < base_version) {
        uint64_t oldest = _versions[0].version;
        if (version < oldest) {
            return Status::NotFound(
                    StringPrintf("version %zu(oldest=%zu) deleted", version, oldest));
        }
        DCHECK_GT(_base_idx, 0);
        for (ssize_t i = static_cast<ssize_t>(_base_idx) - 1; i >= 0; i--) {
            uint64_t v = _versions[i].version;
            if (v >= version) {
                DCHECK(_versions[i].delta);
                *real_version = v;
                deltas->emplace_back(_versions[i].delta.get());
                if (v == version) {
                    break;
                }
            } else {
                break;
            }
        }
    } else if (version > base_version) {
        size_t vsize = _versions.size();
        for (size_t i = _base_idx + 1; i < vsize; i++) {
            uint64_t v = _versions[i].version;
            if (v <= version) {
                DCHECK(_versions[i].delta);
                *real_version = v;
                deltas->emplace_back(_versions[i].delta.get());
                if (v == version) {
                    break;
                }
            } else {
                break;
            }
        }
    }
    return Status::OK();
}

void Column::capture_latest(vector<ColumnDelta*>* deltas, uint64_t* version) const {
    deltas->reserve(_versions.size() - _base_idx - 1);
    for (size_t i = _base_idx + 1; i < _versions.size(); i++) {
        deltas->emplace_back(_versions[i].delta.get());
    }
    *version = _versions.back().version;
}

Status Column::create_reader(uint64_t version, std::unique_ptr<ColumnReader>* reader) {
    ColumnType type = schema().type();
    bool nullable = schema().is_nullable();
    std::vector<ColumnDelta*> deltas;
    uint64_t real_version;
    RETURN_IF_ERROR(capture_version(version, &deltas, &real_version));

#define CREATE_READER(T)                                                                          \
    if (nullable) {                                                                               \
        (*reader).reset(                                                                          \
                new TypedColumnReader<T, true>(this, version, real_version, std::move(deltas)));  \
    } else {                                                                                      \
        (*reader).reset(                                                                          \
                new TypedColumnReader<T, false>(this, version, real_version, std::move(deltas))); \
    }

    switch (type) {
    case OLAP_FIELD_TYPE_BOOL:
    case OLAP_FIELD_TYPE_TINYINT:
        CREATE_READER(int8_t)
        break;
    case OLAP_FIELD_TYPE_SMALLINT:
        CREATE_READER(int16_t)
        break;
    case OLAP_FIELD_TYPE_INT:
        CREATE_READER(int32_t)
        break;
    case OLAP_FIELD_TYPE_BIGINT:
        CREATE_READER(int64_t)
        break;
    case OLAP_FIELD_TYPE_LARGEINT:
        CREATE_READER(int128_t)
        break;
    case OLAP_FIELD_TYPE_FLOAT:
        CREATE_READER(float)
        break;
    case OLAP_FIELD_TYPE_DOUBLE:
        CREATE_READER(double)
        break;
    default:
        return Status::NotSupported("create column reader: type not supported");
    }
#undef CREATE_READER
    return Status::OK();
}

Status Column::create_writer(std::unique_ptr<ColumnWriter>* writer) {
    ColumnType type = schema().type();
    bool nullable = schema().is_nullable();
    std::vector<ColumnDelta*> deltas;
    uint64_t real_version;
    capture_latest(&deltas, &real_version);

#define CREATE_WRITER(T)                                                                  \
    if (nullable) {                                                                       \
        (*writer).reset(new TypedColumnWriter<T, true>(this, real_version, real_version,  \
                                                       std::move(deltas)));               \
    } else {                                                                              \
        (*writer).reset(new TypedColumnWriter<T, false>(this, real_version, real_version, \
                                                        std::move(deltas)));              \
    }

    switch (type) {
    case OLAP_FIELD_TYPE_BOOL:
    case OLAP_FIELD_TYPE_TINYINT:
        CREATE_WRITER(int8_t)
        break;
    case OLAP_FIELD_TYPE_SMALLINT:
        CREATE_WRITER(int16_t)
        break;
    case OLAP_FIELD_TYPE_INT:
        CREATE_WRITER(int32_t)
        break;
    case OLAP_FIELD_TYPE_BIGINT:
        CREATE_WRITER(int64_t)
        break;
    case OLAP_FIELD_TYPE_LARGEINT:
        CREATE_WRITER(int128_t)
        break;
    case OLAP_FIELD_TYPE_FLOAT:
        CREATE_WRITER(float)
        break;
    case OLAP_FIELD_TYPE_DOUBLE:
        CREATE_WRITER(double)
        break;
    default:
        return Status::NotSupported("create column writer: type not supported");
    }
#undef CREATE_WRITER
    return Status::OK();
}

} // namespace memory
} // namespace doris
